[add]: Added related categories in more information section

muskangupta-iitr opened the pull request:
<!--
    Thanks for opening a pull request (PR). Please read through
    these instructions to help us review and merge your change quicker.

    Replace ISSUE_NUMBER with the number of the issue related to what
    you're fixing followed by a description of your change. For example:

    Fixes #3588

    This patch adds a margin on the logout button to correct the layout.
-->

Fixes #2758

Added the related categories in more information section of addon.

<!--
    Read through this checklist to make sure your patch is ready for review.

    - Add a PR title that summarizes your patch
    - Make sure this PR relates to an existing open issue and there are no existing PRs open for the same issue.
    - Add tests to cover the changes added in this PR.
    - Check that the change works locally.
    - Add before and after screenshots (Only for changes that impact the UI).

    Thanks for your contribution!
-->

muskangupta-iitr commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -110,6 +116,8 @@ export class AddonMoreInfoBase extends React.Component<InternalProps> {
       );
     }
 
+    const addonType = addon.type;
```
I will change it as addon variable is coming from `props`.

willdurand commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -173,6 +181,23 @@ export class AddonMoreInfoBase extends React.Component<InternalProps> {
+                  <Button
+                    className={`Categories-link
+                      Categories--category-color-${(index % 12) + 1}`}
+                    to={categoryResultsLinkTo({ addonType, category })}
```
I think we can simplify here as we only need the category names without buttons or colors, see the mockup in this comment: https://github.com/mozilla/addons-frontend/issues/2758#issuecomment-339620743

muskangupta-iitr replied:
Ohk @wildurand. And what about the categories names as only slug is coming ?

muskangupta-iitr replied:
Ohk @willdurand 👍

muskangupta-iitr commented:
@bobsilverberg Have you gone through the PR? I am facing problem in writing its tests.

bobsilverberg commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -9,33 +9,53 @@ import Link from 'amo/components/Link';
+  dispatch: DispatchFunc,
+  errorHandler: ErrorHandlerType,
   hasStatsPermission: boolean,
+  loading: boolean,
```
I would call this `categoriesLoading` to be more specific about what is loading.

muskangupta-iitr replied:
Yes, this would specify loading state.

bobsilverberg commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -9,33 +9,53 @@ import Link from 'amo/components/Link';
 
 type InternalProps = {|
   ...Props,
+  categoriesState: $PropertyType<CategoriesStateType, 'categories'>,
```
Rather than store all the categories from the state in a prop, I think it would simplify this component if we only stored the list of related categories, in a prop called `relatedCategories`, for example. This would mean moving the logic from this component into a function in the `categories` reducer. I will comment on this more below.

bobsilverberg commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -9,33 +9,53 @@ import Link from 'amo/components/Link';
 type InternalProps = {|
   ...Props,
+  categoriesState: $PropertyType<CategoriesStateType, 'categories'>,
+  clientApp: string | null,
```
If we don't store all the categories in a prop, then we probably won't need `clientApp` in `InternalProps`.

bobsilverberg commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -9,33 +9,53 @@ import Link from 'amo/components/Link';
+    super(props);
+    const { categoriesState, dispatch, errorHandler, loading } = props;
+
+    if (!loading && !categoriesState) {
```
We will need logic here to fetch the categories if they are not currently loading and if they haven't already been populated, but rather than look at `categoriesState` you could look at the new prop I suggested, `relatedCategories`.

bobsilverberg commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -54,6 +78,38 @@ export class AddonMoreInfoBase extends React.Component<InternalProps> {
       });
     }
 
+    const addonType = addon.type;
```
I suggest that all of the logic below could be moved into the `categories` reducer. There would be a function called, for example, `getCategoryNames`, which could be passed an array of category `slug`s and would return an array of category names. This function could be called in `mapStateToProps`.

bobsilverberg commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -54,6 +78,38 @@ export class AddonMoreInfoBase extends React.Component<InternalProps> {
+    }
+
+    if (!errorHandler.hasError() && !loading && !categories.length) {
+      return (
```
I don't think you want to just return a `Card` here. Wouldn't that mean that none of the other data would be displayed for the add-on? I suspect that you don't need any special logic for this here at all.

muskangupta-iitr replied:
Yes, you are correct, if we add this than whole addon will not displayed.

bobsilverberg commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -196,14 +256,16 @@ export class AddonMoreInfoBase extends React.Component<InternalProps> {
+    const { addon, i18n, errorHandler } = this.props;
     return (
       <>
+        {errorHandler.renderErrorIfPresent()}
```
I'm not sure if this is the best place to render this error or not. I suggest you take a screenshot of how this looks when the error is rendered and attach it to the PR.

muskangupta-iitr replied:
Ohk, I'll do that!

bobsilverberg commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -326,6 +395,9 @@ export const mapStateToProps = (state: AppState, ownProps: Props) => {
     userId: state.users.currentUserID,
+    clientApp: state.api.clientApp,
+    loading: state.categories.loading,
+    categoriesState: state.categories.categories,
```
Instead of returning `state.categories.categories`, in this `mapStateToProps` function you could call the new function that I discussed above, `getCategoryNames` (which you will create in the `categories` reducer, and return the list of category names in `relatedCategories`.

bobsilverberg commented on src/amo/components/AddonMoreInfo/styles.scss:
```diff
@@ -0,0 +1,3 @@
+.Categories-list--justify-content {
+  justify-content: start;
```
Is this really needed?

muskangupta-iitr replied:
No, for the simple design it is not required. Initially I implemented different design, so added this. I will remove this in the next update.

bobsilverberg commented on src/amo/components/Categories/index.js:
```diff
@@ -32,7 +32,7 @@ type CategoryType = {|
 |};
 
-type CategoriesStateType = {|
+export type CategoriesStateType = {|
```
You won't need this export with the suggestions I made above.

muskangupta-iitr commented:
@bobsilverberg Thanks for reviewing the PR. You suggested me a approach above, Correct me I am wrong, I will have to create a reducer called `fetchRelatedCategories` and call the same the saga `fetchCategories`and store only related categories in `relatedCategories` field. Then I call a function `getCategoriesNames` from component.

bobsilverberg commented:
> @bobsilverberg Thanks for reviewing the PR. You suggested me a approach above, Correct me I am wrong, I will have to create a reducer called `fetchRelatedCategories` and call the same the saga `fetchCategories`and store only related categories in `relatedCategories` field. Then I call a function `getCategoriesNames` from component.

No, I'm not suggesting a new fetch function, just a helper function that can use the data in the `categories` reducer. You'd add code to the [reducer](https://github.com/mozilla/addons-frontend/blob/master/src/core/reducers/categories.js), which would be somewhat similar to [`getAddonByID`](https://github.com/mozilla/addons-frontend/blob/master/src/core/reducers/addons.js#L235-L240) in `addons.js`, for example. 

Your new function, which could be called `getCategoryNames` would take two arguments: the first would be the categories state (see how `getAddonByID` accepts `addons` as its first argument), and the second would be an array of category slugs. Then, inside that function you could have the logic that looks through the state to find the names that correspond to the slugs, and then returns that array of names.

The `AddonMoreInfo` component could use this new `getCategoryNames` function inside `mapStateToProps` to store the names in a prop called `relatedCategories`.

Does that make sense?

muskangupta-iitr commented:
@bobsilverberg Yeah, I get that. Let me update the PR with the changes.

muskangupta-iitr commented:
@bobsilverberg Done!
Here is the screenshot:

![Screenshot from 2020-03-28 03-40-47](https://user-images.githubusercontent.com/50856599/77804772-56b7f580-70a6-11ea-87c9-7af647c7d5c9.png)

bobsilverberg commented on src/core/reducers/categories.js:
```diff
@@ -96,6 +106,47 @@ export function createEmptyCategoryList(): EmptyCategoryListType {
   }, {});
 }
 
+export const getCategories = (
```
I don't think we need `getCategores` at all. It's not used anywhere except to generate data that is then passed into `getCategoryNames`. 

We'll need `getCategoryNames` to also accept arguments for `appName` and `addonType`, but if you do that then I believe you can add the logic of `getCategores` into `getCategoryNames` and then `AddonMoreInfo` will only have to make one call to `getCategoryNames`, and not two calls to `getCategores` and then to `getCategoryNames`.

Does that make sense?

muskangupta-iitr replied:
Yes, that make sense but I found that in `amo/components/Categories/index.js` also `getCategories` can be called  replacing the code
```
    if (
      categoriesState &&
      categoriesState[clientApp] &&
      categoriesState[clientApp][addonType]
    ) {
      categories = Object.values(categoriesState[clientApp][addonType]);
    }
```

muskangupta-iitr replied:
@bobsilverberg That's why I created that `getCategories` function. So, should I remove `getCategories` function?

bobsilverberg replied:
Ah, I didn't realize that because you didn't include changes to `amo/components/Categories/index.js` in this patch, but that's a good find! If that's the case, then it's probably worthwhile having. Still, you could call `getCategories` from `getCategoryNames` in the reducer, thus avoiding calling both `getCategories` and `getCategoryNames` from `AddonMoreInfo`. Also, please include the changes to `amo/components/Categories/index.js` in this patch.

muskangupta-iitr replied:
Yes sure! I will change that too and start working on the tests.

muskangupta-iitr commented:
@bobsilverberg As most of the changes have been approved by you, so can I start working on tests?

bobsilverberg commented:
> @bobsilverberg As most of the changes have been approved by you, so can I start working on tests?

I did make a suggestion about another change, but yes, you can start working on tests at any time.

bobsilverberg commented:
> @bobsilverberg Done!
> Here is the screenshot:
> 
> ![Screenshot from 2020-03-28 03-40-47](https://user-images.githubusercontent.com/50856599/77804772-56b7f580-70a6-11ea-87c9-7af647c7d5c9.png)

I had suggested you add a screenshot of an example with the error message displayed, as discussed at https://github.com/mozilla/addons-frontend/pull/9253#discussion_r399264447.

muskangupta-iitr commented:
> > @bobsilverberg Done!
> > Here is the screenshot:
> > ![Screenshot from 2020-03-28 03-40-47](https://user-images.githubusercontent.com/50856599/77804772-56b7f580-70a6-11ea-87c9-7af647c7d5c9.png)
> 
> I had suggested you add a screenshot of an example with the error message displayed, as discussed at [#9253 (comment)](https://github.com/mozilla/addons-frontend/pull/9253#discussion_r399264447).

Thanks for the reminder :) Just want to ask that Can I remove this line `       {errorHandler.renderErrorIfPresent()}` as if any thing goes wrong with the related categories, then the related categories will not render.

bobsilverberg commented:
> 
> Thanks for the reminder :) Just want to ask that Can I remove this line ` {errorHandler.renderErrorIfPresent()}` as if any thing goes wrong with the related categories, then the related categories will not render.

Yes, I think that makes sense.

muskangupta-iitr commented:
> > Thanks for the reminder :) Just want to ask that Can I remove this line ` {errorHandler.renderErrorIfPresent()}` as if any thing goes wrong with the related categories, then the related categories will not render.
> 
> Yes, I think that makes sense.

Thanks:) I will remove that.

muskangupta-iitr commented:
@bobsilverberg I getting multiple errors in the tests, Could you please help me solving those? I got stuck here....

Here are the screenshots:

![Screenshot from 2020-04-02 02-59-42](https://user-images.githubusercontent.com/50856599/78188595-213c4f00-748e-11ea-8a2a-f8b972a3860d.png)
![Screenshot from 2020-04-02 02-59-38](https://user-images.githubusercontent.com/50856599/78188640-387b3c80-748e-11ea-971a-f91af0bf5e45.png)
![Screenshot from 2020-04-02 02-59-33](https://user-images.githubusercontent.com/50856599/78188693-4b8e0c80-748e-11ea-8dd5-8426167a6e6e.png)
![Screenshot from 2020-04-02 02-59-28](https://user-images.githubusercontent.com/50856599/78188753-652f5400-748e-11ea-973f-1ae67a7feea0.png)
![Screenshot from 2020-04-02 02-59-25](https://user-images.githubusercontent.com/50856599/78188794-76786080-748e-11ea-9aa5-53a36a967632.png)
![Screenshot from 2020-04-02 02-59-22](https://user-images.githubusercontent.com/50856599/78188832-8abc5d80-748e-11ea-9868-a2f06019ad06.png)
![Screenshot from 2020-04-02 02-59-19](https://user-images.githubusercontent.com/50856599/78188878-a45da500-748e-11ea-8bc2-a927ee1762d0.png)

muskangupta-iitr commented:
@bobsilverberg I am able to solve solve some of the error above but still getting some. I found that CategoryType and CategoryStateType in the file src/amo/components/Categories are different from ExternalCategory and CategoryMapType in the src/core/reducers/categories.js due to that I am facing some issues.

Here are the errors:
Error -------------------------------------------------------------------- src/amo/components/Categories/index.js:117:34

Cannot call getCategories with categoriesState bound to categoriesState because:
 - undefined [1] is incompatible with object type [2] in the indexer property.
 - undefined [3] is incompatible with object type [4] in the indexer property of the indexer property.
 - string [5] is incompatible with null [6] in property description of the indexer property of the indexer property of
   the indexer property.
 - number [7] is incompatible with string [8] in property id of the indexer property of the indexer property of the
   indexer property.
 - undefined [5] is incompatible with null [6] in property description of the indexer property of the indexer property
   of the indexer property.

     src/amo/components/Categories/index.js
 [5]  26|   description?: string,
 [7]  27|   id: number,
        :
 [1]  37|     [clientApp: string]: void | {
 [3]  38|       [addonType: string]: void | { [categorySlug: string]: CategoryType },
        :
     114|     let categories = [];
     115|
     116|     if (categoriesState && clientApp && addonType) {
     117|       categories = getCategories(categoriesState, clientApp, addonType);
     118|     }
     119|
     120|     const classNameProp = classnames('Categories', className);

     src/core/reducers/categories.js
 [8]  14|   id: string,
        :
 [6]  21|   description: string | null,
        :
 [2]  27|   [appName: string]: {
 [4]  28|     [addonType: string]: {
      29|       [addonSlug: string]: ExternalCategory,
      30|     },
      31|   },

muskangupta-iitr commented:
@bobsilverberg I have pushed a commit by solving some flow type errors but still facing few. Could you please review this commit? Thanks :smile:

muskangupta-iitr commented:
@bobsilverberg I am stuck here and need help. Could you please review my patch as I unable to work further without assistance? :/

muskangupta-iitr commented:
@bobsilverberg I am stuck here and need help. Could you please review my patch as I am unable to work further without assistance? :/

bobsilverberg commented:
My apologies @muskangupta-iitr. I was busy with some other things late last week, and then Friday was a holiday. I will take a look at this today.

muskangupta-iitr commented:
> I resolved the remaining Flow issues. There was a problem with the code you were using to retrieve the `categories` using `getCategories` in the `render` method, and a nice way of dealing with that was to actually remove that from the `render` method and just retrieve the `categories` in `mapStateToProps`, which is really where it should be done. Here is a diff of the changes I made that did this and fixed the Flow issues:
> 
> ```diff
> diff --git a/src/amo/components/Categories/index.js b/src/amo/components/Categories/index.js
> index 9d91b2db4..adc234488 100644
> --- a/src/amo/components/Categories/index.js
> +++ b/src/amo/components/Categories/index.js
> @@ -48,8 +48,8 @@ type Props = {|
> 
>  type InternalProps = {|
>    ...Props,
> +  categories: Array<string>,
>    categoriesState: $PropertyType<CategoriesStateType, 'categories'>,
> -  clientApp: string,
>    dispatch: DispatchFunc,
>    errorHandler: ErrorHandlerType,
>    i18n: I18nType,
> @@ -102,21 +102,14 @@ export class CategoriesBase extends React.Component<InternalProps> {
>      /* eslint-disable react/no-array-index-key */
>      const {
>        addonType,
> -      categoriesState,
> +      categories,
>        className,
> -      clientApp,
>        errorHandler,
>        i18n,
>        loading,
>      } = this.props;
>      invariant(addonType, 'addonType is undefined');
> 
> -    let categories = [];
> -
> -    if (categoriesState && clientApp && addonType) {
> -      categories = getCategories(categoriesState, clientApp, addonType);
> -    }
> -
>      const classNameProp = classnames('Categories', className);
> 
>      if (!errorHandler.hasError() && !loading && !categories.length) {
> @@ -180,10 +173,19 @@ export class CategoriesBase extends React.Component<InternalProps> {
>    }
>  }
> 
> -export function mapStateToProps(state: AppState) {
> +export function mapStateToProps(state: AppState, ownProps: Props) {
> +  const { addonType } = ownProps;
> +
>    return {
> +    categories:
> +      state.api.clientApp && state.categories.categories
> +        ? getCategories(
> +            state.categories.categories,
> +            state.api.clientApp,
> +            addonType,
> +          )
> +        : [],
>      categoriesState: state.categories.categories,
> -    clientApp: state.api.clientApp,
>      loading: state.categories.loading,
>    };
>  }
> ```

@bobsilverberg Thank you so much! It got worked and cleared all the flow type errors. I have added the changes and updated the patch. Also, I am finding it difficult to pass `relatedCategories` in the test. So, I just added a check for that.
```
relatedCategoriesObj  = ownProps.relatedCategories || null
```
Is it fine?

bobsilverberg commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -291,7 +327,6 @@ export class AddonMoreInfoBase extends React.Component<InternalProps> {
 
   render() {
     const { i18n } = this.props;
-
```
I would put this blank line back in.

muskangupta-iitr replied:
Sure! I will put that back in.

bobsilverberg commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -301,7 +336,21 @@ export class AddonMoreInfoBase extends React.Component<InternalProps> {
 
 export const mapStateToProps = (state: AppState, ownProps: Props) => {
-  const { addon, i18n } = ownProps;
+  const { addon, i18n, relatedCategories } = ownProps;
```
This should not include `relatedCategories`.

bobsilverberg commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -301,7 +336,21 @@ export class AddonMoreInfoBase extends React.Component<InternalProps> {
+  const categoriesState = state.categories.categories;
+  const appName = state.api.clientApp;
+
+  let relatedCategoriesObj = relatedCategories || null;
```
I think you said you added this to help a test pass, but you shouldn't need it.

muskangupta-iitr replied:
Yes, I need that because when I am running the test, I am always getting relatedCategories, after passing relatedCategories from tests, as null because in `mapStateToProps` as I initialized it to null. That's why I added this check.

bobsilverberg replied:
That just means you need a different set up in the test. We shouldn't add code that we don't need just to get a test to pass - that's a sign there's something wrong with the test.

muskangupta-iitr replied:
Ohk! I will change that.

bobsilverberg commented on src/amo/components/Categories/index.js:
```diff
@@ -48,8 +48,8 @@ type Props = {|
 type InternalProps = {|
   ...Props,
+  categories: Array<string>,
   categoriesState: $PropertyType<CategoriesStateType, 'categories'>,
```
You don't need this as an InternalProp anymore. It can also be removed from `mapStateToProps`.

muskangupta-iitr replied:
@bobsilverberg Sorry, for not replying on this question. But if I remove the categories from `mapStateToProps`, how will I get the `categories` in the component ?
Also, I want to ask that why we are removing it from `InternalProps`?  I found some error in here, instead of `Array<string>` it should `Array<CategoryType>`?

bobsilverberg commented on src/core/reducers/categories.js:
```diff
@@ -96,6 +104,47 @@ export function createEmptyCategoryList(): EmptyCategoryListType {
+  categoriesState: CategoryMapType,
+  appName: string,
+  addonType: string,
+): Array<mixed> | null => {
```
This shouldn't be `Array<mixed>`, it should be an Array of the type you are returning.

muskangupta-iitr replied:
It should be string but when I put string in here, I am getting a flow type error 
```
Cannot return `relatedCategories` because  mixed [1] is incompatible with  string [2] in array element.
```
For removing this I used mixed.

bobsilverberg replied:
As mentioned above, create a new commit with all the changes I suggested (and use `string` here, like you were originally), and then I can help you debug any Flow errors you have. Flow can be quite tricky, especially when you're first getting used to it.

bobsilverberg replied:
Looking at this some more, this might also be the reason that you needed `mixed` in `getCategoryNames`. I think there's something wrong with this `getCategories` function.

muskangupta-iitr replied:
I think instead of using Object.values directly, I can use Object.keys as I read one of the github issue about this, also mentioned in one of the TODO [https://github.com/facebook/flow/issues/2221](https://github.com/facebook/flow/issues/2221)

bobsilverberg commented on src/core/reducers/categories.js:
```diff
@@ -96,6 +104,47 @@ export function createEmptyCategoryList(): EmptyCategoryListType {
+  addonType: string,
+): Array<mixed> | null => {
+  if (
+    categoriesState &&
```
I don't think this part of the condition is needed. The function says that `categoriesState` is required, so it must exist.

bobsilverberg commented on src/core/reducers/categories.js:
```diff
@@ -31,6 +31,14 @@ type CategoryMapType = {
+  [addonSlug: string]: ExternalCategory,
+};
+
+type AddonCategories = {
```
I don't think we need this. It's nice that it's specific about what `AddonCategories` looks like, but the type of `categories` for an addon is already defined in `AddonType`. So instead of this you can just reference that via `$PropertyType<AddonType, 'categories'>`.

muskangupta-iitr replied:
Ohk! I don't know about this. Good to know this type of variable type checking. I will update that.

bobsilverberg commented on src/core/reducers/categories.js:
```diff
@@ -31,6 +31,14 @@ type CategoryMapType = {
   },
 };
 
+export type CategorySlugType = {
```
This doesn't appear to be used anywhere.

muskangupta-iitr replied:
Yes, I am testing some type variable with this, but just forgot to remove. I will update this.

bobsilverberg commented on src/core/reducers/categories.js:
```diff
@@ -96,6 +104,47 @@ export function createEmptyCategoryList(): EmptyCategoryListType {
+
+export function getCategoryNames(
+  categoriesState: CategoryMapType,
+  addonCategories: AddonCategories,
```
As mentioned above, in place of `AddonCategories`, you can use `$PropertyType<AddonType, 'categories'>`

muskangupta-iitr replied:
Yes, I will update that.

bobsilverberg commented on src/core/reducers/categories.js:
```diff
@@ -96,6 +104,47 @@ export function createEmptyCategoryList(): EmptyCategoryListType {
+  addonCategories: AddonCategories,
+  appName: string,
+  addonType: string,
+): Array<mixed> | null {
```
As above, this shouldn't be `Array<mixed>`, it should be an Array of the type you are returning.

muskangupta-iitr replied:
It should be string but when I put string in here, I am getting a flow type error 
```
Cannot return `relatedCategories` because  mixed [1] is incompatible with  string [2] in array element.
```
For removing this I used mixed.

bobsilverberg replied:
As mentioned above, try changing it back to `string`, and push a new commit with all of the changes, and I can help you debug the Flow errors.

muskangupta-iitr replied:
Ohk!

bobsilverberg commented on src/core/reducers/categories.js:
```diff
@@ -96,6 +104,47 @@ export function createEmptyCategoryList(): EmptyCategoryListType {
+  appName: string,
+  addonType: string,
+): Array<mixed> | null => {
+  if (
```
I guess we need `invariants` for all of the required arguments like we have in `getCategoryNames` below.

muskangupta-iitr replied:
Instead of `if` we have to use `invariants` like in this way:
```
  invariant(categoriesState, 'categories state can not be empty');
  invariant(categoriesState[appName], 'categories state does not contains the app name');
  invariant(categoriesState[appName][addonType], 'categories state does not contain the addon type');

  return Object.values(categoriesState[appName][addonType]);
```
Am I correct ?

bobsilverberg replied:
No. We should just have invariants for `categoriesState`, `appName` and `addonType`, because those are the required arguments. You can then keep the logic you have in the `if` statement for `categoriesState[appName] && categoriesState[appName][addonType]` because there is still a possibility that those will be empty/null.

muskangupta-iitr replied:
Oh correct! I will change this accordingly.

bobsilverberg commented on src/core/reducers/categories.js:
```diff
@@ -96,6 +104,47 @@ export function createEmptyCategoryList(): EmptyCategoryListType {
+  appName: string,
+  addonType: string,
+): Array<mixed> | null {
+  invariant(categoriesState, 'categories state can not be empty!');
```
The exclamation point at the end is unnecessary. Same goes for the other `invariant` calls.

muskangupta-iitr replied:
Ohk! I will remove the exclamations.

bobsilverberg commented on src/core/reducers/categories.js:
```diff
@@ -96,6 +104,47 @@ export function createEmptyCategoryList(): EmptyCategoryListType {
+  addonType: string,
+): Array<mixed> | null {
+  invariant(categoriesState, 'categories state can not be empty!');
+  invariant(addonCategories, 'slugs can not be empty!');
```
As the argument is called `addonCategories`, this message should probably say `addon categories` and not `slugs`.

muskangupta-iitr replied:
Yes, I am also little confused about what should be the string. But, yes it should be on variable, I will update that.

bobsilverberg commented on src/core/reducers/categories.js:
```diff
@@ -96,6 +104,47 @@ export function createEmptyCategoryList(): EmptyCategoryListType {
+  const relatedCategories = [];
+
+  if (categories && addonCategories && addonCategories[appName]) {
+    categories.forEach((r) => {
```
We tend not to use short variables, like `r`, even in a loop. This might be better as `category`.

muskangupta-iitr replied:
I will replace `r` with `category`.

muskangupta-iitr commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -9,33 +9,55 @@ import Link from 'amo/components/Link';
 type Props = {|
   addon: AddonType | null,
   i18n: I18nType,
+  relatedCategories: Array<string> | null,
```
Sure! I will update this.

bobsilverberg replied:
No, it should definitely be in `InternalProps`, just like the other props that are being added via `mapStateToProps`, e.g., `currentVersion`. If you're getting a Flow error it might be due to some of the other changes I suggest you make around `relatedCategories`. The best thing is to make all of the changes I suggested, and if you're still getting Flow errors push a new commit and I can help you debug them.

muskangupta-iitr replied:
Ohk! I will create a commit with the change.

muskangupta-iitr commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -173,6 +196,10 @@ export class AddonMoreInfoBase extends React.Component<InternalProps> {
       ) : null,
+      relatedCategories:
+        relatedCategories && relatedCategories.length > 0
+          ? i18n.gettext(relatedCategories.join(', '))
```
So, it's just `relatedCategories.join(', ')`. Am I correct @bobsilverberg ?

bobsilverberg replied:
Correct

muskangupta-iitr replied:
Thanks, I will update this.

muskangupta-iitr commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -173,6 +196,10 @@ export class AddonMoreInfoBase extends React.Component<InternalProps> {
         </Link>
       ) : null,
+      relatedCategories:
+        relatedCategories && relatedCategories.length > 0
```
Yes, I will update that!

bobsilverberg commented:
Thanks for all your comments and questions @muskangupta-iitr. I have replied to all of them.

muskangupta-iitr commented:
> Thanks for all your comments and questions @muskangupta-iitr. I have replied to all of them.

Thanks for the time and efforts. I am sorry for not commenting and working yesterday; was not well enough to work out the code. I will push a commit with the suggested changes asap. Thanks :)

muskangupta-iitr commented:
@bobsilverberg Thanks for the above suggestion. I have updated the patch with the above changes and I am not getting any flow errors. I am just getting a single error in the test i.e 
```
    toHaveText must be called on a single node, not an empty node.

      565 |     expect(
      566 |       root.find('.AddonMoreInfo-related-categories').children(),
    > 567 |     ).toHaveText('Testing category, Another testing category');
          |       ^
      568 |   });
      569 | 
      570 |   it('links to version history if add-on is extension', () => {

      at Object.<anonymous> (tests/unit/amo/components/TestAddonMoreInfo.js:567:7)

``` 
For this error only, I put that relatedCategories check in `mapStateToProps`. I have to pass relatedCategories from the tests to component but as relatedCategories is redeclared in the `mapStateToProps` as `null`, that is the reason related categories is not rendering. I am thinking of writing an action to set related categories as we do with the versions but I am not sure. Any suggestions on this ?

bobsilverberg commented on tests/unit/amo/components/TestAddonMoreInfo.js:
```diff
@@ -512,6 +515,58 @@ describe(__filename, () => {
+      },
+    });
+
+    const fakeCategoryState = {
```
The approach you are taking here won't work - we need to do something quite different. What we need to do is create data in the redux store that will support our test case. That way, when `mapStateToProps` is executed in `AddonMoreInfo`, it will yield the expected result. This is a bit complex, but I will include a test below that works. We need to do the following things:

1. Create an add-on with the categories we expect, for a specific `clientApp`. You are already doing this, but my code is a bit different.
2. Load the associated categories into the redux state. You are attempting to do something similar here, but, as I said, this won't work. You need to use `loadCategories` instead.
3. Make sure that the `clientApp` for the request matches the one you used for the categories in the `addon`, by using `dispatchClientMetadata`.

Applying these changes to your patch will yield the test I am talking about:
```diff
diff --git a/tests/unit/amo/components/TestAddonMoreInfo.js b/tests/unit/amo/components/TestAddonMoreInfo.js
index 7ca18f864..98bbcfe96 100644
--- a/tests/unit/amo/components/TestAddonMoreInfo.js
+++ b/tests/unit/amo/components/TestAddonMoreInfo.js
@@ -12,7 +12,7 @@ import {
   CLIENT_APP_FIREFOX,
 } from 'core/constants';
 import { createInternalAddon } from 'core/reducers/addons';
-import { getCategoryNames } from 'core/reducers/categories';
+import { loadCategories } from 'core/reducers/categories';
 import { formatFilesize } from 'core/i18n/utils';
 import {
   createContextWithFakeRouter,
@@ -21,6 +21,7 @@ import {
   dispatchClientMetadata,
   dispatchSignInActions,
   fakeAddon,
+  fakeCategory,
   fakeTheme,
   fakeI18n,
   fakeVersion,
@@ -515,56 +519,39 @@ describe(__filename, () => {
     expect(statsLink).toHaveLength(1);
   });

-  it('render related categories of the addon if existed', () => {
+  it('renders related categories of the addon', () => {
+    const category1 = {
+      ...fakeCategory,
+      type: ADDON_TYPE_EXTENSION,
+      name: 'Category 1',
+      slug: 'category-1',
+    };
+    const category2 = {
+      ...fakeCategory,
+      type: ADDON_TYPE_EXTENSION,
+      name: 'Category 2',
+      slug: 'category-2',
+    };
+
     const addon = createInternalAddon({
       ...fakeAddon,
+      type: ADDON_TYPE_EXTENSION,
       categories: {
-        firefox: ['other', 'testing'],
+        [CLIENT_APP_FIREFOX]: [category1.slug, category2.slug],
       },
     });

-    const fakeCategoryState = {
-      firefox: {
-        extension: {
-          other: {
-            id: 1,
-            name: 'Testing category',
-            slug: 'other',
-            application: CLIENT_APP_FIREFOX,
-            misc: false,
-            type: ADDON_TYPE_EXTENSION,
-            weight: 1,
-            description: 'I am a cool category for doing things',
-          },
-          testing: {
-            id: 2,
-            name: 'Another testing category',
-            slug: 'testing',
-            application: CLIENT_APP_FIREFOX,
-            misc: false,
-            type: ADDON_TYPE_EXTENSION,
-            weight: 1,
-            description: 'I am a cool category for doing things',
-          },
-        },
-      },
-    };
-
-    const relatedCategories = getCategoryNames(
-      fakeCategoryState,
-      addon.categories,
-      CLIENT_APP_FIREFOX,
-      addon.type,
-    );
+    dispatchClientMetadata({ clientApp: CLIENT_APP_FIREFOX, store });
+    store.dispatch(loadCategories({ results: [category1, category2] }));

     const root = render({
       addon,
-      relatedCategories,
+      store,
     });

     expect(
       root.find('.AddonMoreInfo-related-categories').children(),
-    ).toHaveText('Testing category, Another testing category');
+    ).toHaveText(`${category1.name}, ${category2.name}`);
   });

   it('links to version history if add-on is extension', () => {
```

muskangupta-iitr replied:
@bobsilverberg Thanks for the above suggestions. It worked perfectly. Actually, I was trying to test that `getCategoryNames` function too in the same test, but yes it should be in the different test and `categories` should dispatched to the store for this.

muskangupta-iitr commented on src/amo/components/AddonMoreInfo/index.js:
```diff
@@ -302,6 +337,20 @@ export class AddonMoreInfoBase extends React.Component<InternalProps> {
+  const categoriesState = state.categories.categories;
+  const appName = state.api.clientApp;
+
+  let relatedCategoriesObj = null;
```
Ohk! I will change that.

muskangupta-iitr commented on src/core/reducers/categories.js:
```diff
@@ -96,6 +97,54 @@ export function createEmptyCategoryList(): EmptyCategoryListType {
+
+  const relatedCategories = [];
+
+  if (categories && addonCategories && addonCategories[appName]) {
```
Yes, i will update this check condition.

muskangupta-iitr commented:
> Thanks for your continued efforts @muskangupta-iitr, this is looking good. There were some issues with your test for `AddonMoreInfo`, but I have made suggestions and showed an example test. You also need to add tests for any other changes you have made, which includes tests for the two new functions in `categories.js`, and we should have a number of other tests for `AddonMoreInfo`, including tests that show that it dispatches `fetchCategories` when needed, and doesn't dispatch `fetchCategories` when not needed, as well as a test that shows that it displays nothing when there are no categories for the `addon`.
> 
> Perhaps you can work on those tests and submit a new commit when you have something. Even if it does not quite work, I can help you with the tests once I can see what you've tried. For the tests for `fetchCategories`, take a look at https://github.com/mozilla/addons-frontend/blob/master/tests/unit/amo/components/TestCategories.js for some inspiration.

@bobsilverberg Thanks for the review and appreciation. It means lot to me. I will add these test cases mentioned in the next patch update. I will try to break it two commits, one fixing current issues and one with new test cases added. Thanks once again :smile:

muskangupta-iitr commented:
@bobsilverberg Where should I add tests for `getCategories` and `getCategoriesNames`, either in `core/reducers/test_categories` or `amo/components/TestAddonMoreInfo.js` or `amo/components/TestCategories.js` ?

muskangupta-iitr commented:
@bobsilverberg Gentle reminder in case you missed my comment, I am stuck and need your help to proceed further. :)

bobsilverberg commented:
> @bobsilverberg Where should I add tests for `getCategories` and `getCategoriesNames` functions, either in `core/reducers/test_categories` or `amo/components/TestAddonMoreInfo.js` or `amo/components/TestCategories.js` ?

`getCategories` and `getCategoriesNames` are in `core/reducers/categories.js`, therefore the tests for them should be in `core/reducers/test_categories`. You will also need to have test coverage for anywhere that you call these new functions, so you will also need to make some changes to tests in `amo/components/TestAddonMoreInfo.js` or `amo/components/TestCategories.js`, but the tests for the new functions themselves should be in `core/reducers/test_categories`.

bobsilverberg commented:
Hi @muskangupta-iitr, just wondering if you're still working on this?

muskangupta-iitr commented:
@bobsilverberg Sorry Bob for being a little unresponsive for past few days and not working on it; got busy with some other work. I still wish to work on this PR and contribute and would be glad if you allow me to do so. :)

bobsilverberg commented:
Of course, we would be happy for you to continue to work on these! Whenever you have the time. I will leave them open for you.

bobsilverberg commented:
@muskangupta-iitr Just checking in to see if you're still planning on working on these PRs. If you don't think you'll get back to them soon I'm going to close them.
